sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/xml/XPathExpressionSuite.scala

```diff
@@ -27,33 +27,127 @@ import org.apache.spark.sql.types.StringType
+      "sum(a/b[@class=\"odd\"])",
+      5.toShort)
+  }
+
+  test("xpath_int") {
+    def testExpr[T](xml: String, path: String, expected: java.lang.Integer): Unit = {
+      checkEvaluation(
+        XPathInt(Literal.create(xml, StringType), Literal.create(path, StringType)),
+        expected)
+    }
+
+    testExpr("<a>this is not a number</a>", "a", 0)
+    testExpr("<a>try a boolean</a>", "a = 10", 0)
+    testExpr("<a><b class=\"odd\">1</b><b class=\"even\">2</b><b class=\"odd\">4</b><c>8</c></a>",
+      "sum(a/b[@class=\"odd\"])",
+      5)
+  }
+
+  test("xpath_long") {
+    def testExpr[T](xml: String, path: String, expected: java.lang.Long): Unit = {
+      checkEvaluation(
+        XPathLong(Literal.create(xml, StringType), Literal.create(path, StringType)),
+        expected)
+    }
+
+    testExpr("<a>this is not a number</a>", "a", 0L)
+    testExpr("<a>try a boolean</a>", "a = 10", 0L)
+    testExpr("<a><b class=\"odd\">1</b><b class=\"even\">2</b><b class=\"odd\">4</b><c>8</c></a>",
+      "sum(a/b[@class=\"odd\"])",
+      5L)
+  }
+
+  test("xpath_float") {
+    def testExpr[T](xml: String, path: String, expected: java.lang.Float): Unit = {
+      checkEvaluation(
+        XPathFloat(Literal.create(xml, StringType), Literal.create(path, StringType)),
+        expected)
+    }
+
+    testExpr("<a>this is not a number</a>", "a", Float.NaN)
+    testExpr("<a>try a boolean</a>", "a = 10", 0.0F)
+    testExpr("<a><b class=\"odd\">1</b><b class=\"even\">2</b><b class=\"odd\">4</b><c>8</c></a>",
+      "sum(a/b[@class=\"odd\"])",
+      5.0F)
+  }
+
+  test("xpath_double") {
+    def testExpr[T](xml: String, path: String, expected: java.lang.Double): Unit = {
+      checkEvaluation(
+        XPathDouble(Literal.create(xml, StringType), Literal.create(path, StringType)),
+        expected)
+    }
+
+    testExpr("<a>this is not a number</a>", "a", Double.NaN)
+    testExpr("<a>try a boolean</a>", "a = 10", 0.0)
+    testExpr("<a><b class=\"odd\">1</b><b class=\"even\">2</b><b class=\"odd\">4</b><c>8</c></a>",
+      "sum(a/b[@class=\"odd\"])",
+      5.0)
+  }
+
+  test("xpath_string") {
+    def testExpr[T](xml: String, path: String, expected: String): Unit = {
+      checkEvaluation(
+        XPathString(Literal.create(xml, StringType), Literal.create(path, StringType)),
+        expected)
+    }
+
+    testExpr("<a><b>bb</b><c>cc</c></a>", "a", "bbcc")
+    testExpr("<a><b>bb</b><c>cc</c></a>", "a/b", "bb")
+    testExpr("<a><b>bb</b><c>cc</c></a>", "a/c", "cc")
+    testExpr("<a><b>bb</b><c>cc</c></a>", "a/d", "")
+    testExpr("<a><b>b1</b><b>b2</b></a>", "//b", "b1")
+    testExpr("<a><b>b1</b><b>b2</b></a>", "a/b[1]", "b1")
+    testExpr("<a><b>b1</b><b id='b_2'>b2</b></a>", "a/b[@id='b_2']", "b2")
+  }
+
+  test("null handling") {
+    // We only do this for one expression since they all share the same common base implementation
```

Comment:
seems the boolean and string one doesn't share same implementation?(`xpathUtil.evalBoolean` and `xpathUtil.evalString`)